File: 3Sum.rb
require 'set'
require 'byebug'
# def three_sum(nums)
#   i = 0
    # result = []
#   while i< nums.length-2
#     j = i+1
#     while j < nums.length -1
#       z = j+1
#       while z < nums.length
#         if nums[i]+nums[j]+nums[z] == 0
#           result << [nums[i], nums[j], nums[z]]
#         end
#         z += 1
#       end
#       j += 1

#     end
#     i += 1

#   end
#   result
    
# end

def three_sum(nums)
  hash = Hash.new{|k,v| k[v]=[]}
  nums.each_with_index do |num, idx|
    hash[num] << idx
  end
  result = []
  i = 0

  while i< nums.length-1
    j = i+1
    while j < nums.length
      dif = -(nums[i]+nums[j])
      if hash[dif] != [] && hash[dif] != [i] && hash[dif] != [j] && hash[dif] != [i, j]
        arr = [nums[i], nums[j], dif].sort
        if !result.include?(arr)
          result.push(arr)
        end
      end
      j += 1
    end
    i+=1
  end
  result
end
  
    
p three_sum([-1,0,1,2,-1,-4])

File: 3Sum.py
def three_sum(nums):
  result = []
  seen_already = set()
  for i in range(len(nums)):
    num = nums[i]
    if not num in seen_already:
      seen_already.add(num)
      arr = []
      if i == len(nums)-1:
        arr = nums[0:-1]
      else: 
        arr = nums[0:i]+nums [i+1:]
      valid_combos = two_sum(arr, -num)
      for combo in valid_combos:
        combo.append(num)
        combo.sort()
      for j in range(len(valid_combos)):
        combo = valid_combos[j]
        if not combo in result:
          result.append(combo)

  return result


def two_sum(nums, target):
  s = set()
  result = []
  for i in range(len(nums)):
    num = nums[i]
    if target-num in s:
      result.append([num, target-num])
    else:
      s.add(num)
      
  return result

print( three_sum([-1,0,1,2,-1,-4]) )
